File: src/main/java/io/myosotisdev/utopianism/listeners/EntityEvents.java
package io.myosotisdev.utopianism.listeners;

import io.myosotisdev.utopianism.battle.DamageMeta;
import net.minestom.server.MinecraftServer;
import net.minestom.server.entity.Entity;
import net.minestom.server.entity.LivingEntity;
import net.minestom.server.event.*;
import net.minestom.server.event.entity.EntityAttackEvent;
import net.minestom.server.event.entity.EntityDamageEvent;
import net.minestom.server.event.entity.EntityDeathEvent;
import net.minestom.server.event.trait.EntityEvent;

import java.util.function.Consumer;

public class EntityEvents
{
    private EventNode<EntityEvent> node;
    private EventNode<EntityEvent> node_LivingEntity;
    Consumer<EntityAttackEvent> consumer_EntityAttackEvent;
    Consumer<EntityDeathEvent> consumer_EntityDeathEvent;

    public EntityEvents()
    {
        MinecraftServer.getGlobalEventHandler()
                       .addChild(node = EventNode.type("entity-events", EventFilter.ENTITY));

        node_LivingEntity = EventNode.event("living-entity-events", EventFilter.ENTITY, entityEvent -> entityEvent.getEntity() instanceof LivingEntity);

        consumer_EntityAttackEvent = event -> {
            LivingEntity victim = (LivingEntity) event.getEntity();

            DamageMeta damageMeta = new DamageMeta((LivingEntity) event.getEntity(), (LivingEntity) event.getTarget());
        };
        node.addListener(EntityAttackEvent.class, consumer_EntityAttackEvent);

        consumer_EntityDeathEvent = event -> {

        };
    }
}

File: src/main/kotlin/io/myosotisdev/utopianism/listeners/EntityEvents.kt
package io.myosotisdev.utopianism.listeners

import io.myosotisdev.utopianism.battle.DamageMetadata
import net.minestom.server.MinecraftServer
import net.minestom.server.entity.LivingEntity
import net.minestom.server.event.EventFilter
import net.minestom.server.event.EventNode
import net.minestom.server.event.entity.EntityAttackEvent
import net.minestom.server.event.entity.EntityDeathEvent
import net.minestom.server.event.entity.projectile.ProjectileCollideWithBlockEvent
import net.minestom.server.event.entity.projectile.ProjectileCollideWithEntityEvent
import net.minestom.server.event.trait.EntityEvent
import java.util.function.Consumer

class EntityEvents
{
    private var node: EventNode<EntityEvent>
    val consumer_EntityAttackEvent: Consumer<EntityAttackEvent> = Consumer { event: EntityAttackEvent ->
        val victim = event.entity as LivingEntity
        val damageMetadata = DamageMetadata(event)
    }
    val consumer_EntityDeathEvent: Consumer<EntityDeathEvent> = Consumer {

    }
    val consumer_ProjectileHitEntityEvent: Consumer<ProjectileCollideWithEntityEvent> = Consumer { event ->

    }
    val consumer_ProjectileHitGroundEvent: Consumer<ProjectileCollideWithBlockEvent> = Consumer { event ->

    }

    init
    {
        MinecraftServer.getGlobalEventHandler()
                .addChild(EventNode.type("entity-events", EventFilter.ENTITY)
                        .also { node = it })

        node.addListener(EntityAttackEvent::class.java, consumer_EntityAttackEvent)
        node.addListener(EntityDeathEvent::class.java, consumer_EntityDeathEvent)
        node.addListener(ProjectileCollideWithEntityEvent::class.java, consumer_ProjectileHitEntityEvent)
        node.addListener(ProjectileCollideWithBlockEvent::class.java, consumer_ProjectileHitGroundEvent)

    }
}